pkg/reconciler/pipelinerun/resources/apply.go

```diff
@@ -124,3 +127,73 @@ func replaceParamValues(params []v1beta1.Param, stringReplacements map[string]st
 	}
 	return params
 }
+
+// ApplyTaskResultsToPipelineResults applies the results of completed TasksRuns to a Pipeline's
+// list of PipelineResults, returning the computed set of PipelineRunResults. References to
+// non-existent TaskResults or failed TaskRuns result in a PipelineResult being considered invalid
+// and omitted from the returned slice. A nil slice is returned if no results are passed in or all
+// results are invalid.
+func ApplyTaskResultsToPipelineResults(results []v1beta1.PipelineResult, taskRunStatuses map[string]*v1beta1.PipelineRunTaskRunStatus) []v1beta1.PipelineRunResult {
+	taskStatuses := map[string]*v1beta1.PipelineRunTaskRunStatus{}
+	for _, trStatus := range taskRunStatuses {
+		taskStatuses[trStatus.PipelineTaskName] = trStatus
+	}
+
+	var runResults []v1beta1.PipelineRunResult = nil
+	stringReplacements := map[string]string{}
+	for _, pipelineResult := range results {
+		variablesInPipelineResult, _ := v1beta1.GetVarSubstitutionExpressionsForPipelineResult(pipelineResult)
+		validPipelineResult := true
+		for _, variable := range variablesInPipelineResult {
+			if _, isMemoized := stringReplacements[variable]; isMemoized {
+				continue
+			}
+			if resultValue := taskResultValue(variable, taskStatuses); resultValue != nil {
+				stringReplacements[variable] = *resultValue
+			} else {
+				validPipelineResult = false
+			}
+		}
+		if validPipelineResult {
+			finalValue := pipelineResult.Value
+			for variable, value := range stringReplacements {
+				v := fmt.Sprintf("$(%s)", variable)
+				finalValue = strings.Replace(finalValue, v, value, -1)
+			}
+			runResults = append(runResults, v1beta1.PipelineRunResult{
+				Name:  pipelineResult.Name,
+				Value: finalValue,
+			})
+		}
+	}
+
+	return runResults
+}
+
+// taskResultValue returns a pointer to the result value for a given task result variable. A nil
+// pointer is returned if the variable is invalid for any reason.
+func taskResultValue(variable string, taskStatuses map[string]*v1beta1.PipelineRunTaskRunStatus) *string {
+	variableParts := strings.Split(variable, ".")
+	if len(variableParts) != 4 || variableParts[0] != "tasks" || variableParts[2] != "results" {
```

Comment:
NIT: `tasks` and `results` could be constants and shared with other result related logic (not part of this PR), can be done seperately